drivers/misc/optee.c

```diff
@@ -0,0 +1,648 @@
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  ret = optee_send_recv(psocket, msg);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  ret = optee_from_msg_param(arg->params, arg->num_params,
+                                msg->params + 2);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  arg->session = msg->session;
+  arg->ret = msg->ret;
+  arg->ret_origin = msg->ret_origin;
+
+  return ret;
+}
+
+static int optee_ioctl_invoke(FAR struct socket *psocket,
+                              FAR struct tee_ioctl_buf_data *buf)
+{
+  char msg_buf[OPTEE_MSG_GET_ARG_SIZE(OPTEE_MAX_PARAM_NUM)];
+  FAR struct tee_ioctl_invoke_arg *arg;
+  FAR struct optee_msg_arg *msg;
+  int ret;
+
+  if (buf->buf_len > TEE_MAX_ARG_SIZE ||
+      buf->buf_len < sizeof(struct tee_ioctl_invoke_arg))
+    {
+      return -EINVAL;
+    }
+
+  arg = (FAR struct tee_ioctl_invoke_arg *)(uintptr_t)buf->buf_ptr;
+
+  if (sizeof(*arg) + TEE_IOCTL_PARAM_SIZE(arg->num_params) !=
+      buf->buf_len)
+    {
+      return -EINVAL;
+    }
+
+  if (arg->num_params > OPTEE_MAX_PARAM_NUM)
+    {
+      return -EINVAL;
+    }
+
+  arg->ret = TEE_ERROR_COMMUNICATION;
+  arg->ret_origin = TEE_ORIGIN_COMMS;
+
+  memset(msg_buf, 0, sizeof(msg_buf));
+  msg = (FAR struct optee_msg_arg *)&msg_buf[0];
+
+  msg->cmd = OPTEE_MSG_CMD_INVOKE_COMMAND;
+  msg->func = arg->func;
+  msg->session = arg->session;
+  msg->cancel_id = arg->cancel_id;
+  msg->num_params = arg->num_params;
+
+  ret = optee_to_msg_param(msg->params, arg->num_params, arg->params);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  ret = optee_send_recv(psocket, msg);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  ret = optee_from_msg_param(arg->params, arg->num_params, msg->params);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  arg->ret = msg->ret;
+  arg->ret_origin = msg->ret_origin;
+
+  return ret;
+}
+
+static int
+optee_ioctl_close_session(FAR struct socket *psocket,
+                          FAR struct tee_ioctl_close_session_arg *arg)
+{
+  struct optee_msg_arg msg;
+
+  memset(&msg, 0, sizeof(struct optee_msg_arg));
+  msg.cmd = OPTEE_MSG_CMD_CLOSE_SESSION;
+  msg.session = arg->session;
+  msg.num_params = 0;
+
+  return optee_send_recv(psocket, &msg);
+}
+
+static int optee_ioctl_version(FAR struct tee_ioctl_version_data *vers)
+{
+  vers->impl_id = TEE_IMPL_ID_OPTEE;
+  vers->impl_caps = TEE_OPTEE_CAP_TZ;
+  vers->gen_caps = TEE_GEN_CAP_GP | TEE_GEN_CAP_MEMREF_NULL;
+  return 0;
+}
+
+static int optee_ioctl_cancel(FAR struct socket *psocket,
+                              FAR struct tee_ioctl_cancel_arg *arg)
+{
+  struct optee_msg_arg msg;
+
+  memset(&msg, 0, sizeof(struct optee_msg_arg));
+  msg.cmd = OPTEE_MSG_CMD_CANCEL;
+  msg.session = arg->session;
+  msg.cancel_id = arg->cancel_id;
+  return optee_send_recv(psocket, &msg);
+}
+
+static int
+optee_ioctl_shm_alloc(FAR struct tee_ioctl_shm_alloc_data *data)
+{
+  int memfd = memfd_create(OPTEE_SERVER_PATH, O_CREAT);
```

Comment:
I have a mixed feeling about this driver. It seems to be registered from the kernel, but uses user space API and access `errno`. Maybe be acceptable for FLAT builds, but not sure about the kernel separated mode.

Reply:
> I have a mixed feeling about this driver. It seems to be registered from the kernel, but uses user space API and access `errno`. Maybe be acceptable for FLAT builds, but not sure about the kernel separated mode.

This should be tested. I have the same concerns. Drivers should not depend on build mode (FLAT, PROTECTED, KERNEL).

Reply:
it's work for protected and kernel mode too, since optee_ioctl_shm_alloc is only invoked from userspace thread, neverr kernel thread.